sqld/src/database/dump_loader.rs

```diff
@@ -0,0 +1,91 @@
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::PathBuf;
+
+use anyhow::anyhow;
+use sqld_libsql_bindings::wal_hook::WalHook;
+use tokio::sync::{mpsc, oneshot};
+
+use super::libsql::open_db;
+
+type OpMsg = Box<dyn FnOnce(&rusqlite::Connection) + 'static + Send + Sync>;
+
+pub struct DumpLoader {
+    sender: mpsc::Sender<OpMsg>,
+}
+
+impl DumpLoader {
+    pub async fn new(
+        path: PathBuf,
+        hooks: impl WalHook + Clone + Send + 'static,
+    ) -> anyhow::Result<Self> {
+        let (sender, mut receiver) = mpsc::channel::<OpMsg>(1);
+
+        let (ok_snd, ok_rcv) = oneshot::channel::<anyhow::Result<()>>();
+        tokio::task::spawn_blocking(move || {
+            let db = match open_db(path, hooks, false) {
+                Ok(db) => {
+                    let _ = ok_snd.send(Ok(()));
+                    db
+                }
+                Err(e) => {
+                    let _ = ok_snd.send(Err(e));
+                    return;
+                }
+            };
+            while let Some(f) = receiver.blocking_recv() {
+                f(&db);
+            }
+        });
+
+        ok_rcv.await??;
+
+        Ok(Self { sender })
+    }
+
+    /// Attempts to load the dump at `path` into the database.
+    pub async fn load_dump(&self, path: PathBuf) -> anyhow::Result<()> {
+        tracing::info!("loading dump at `{}`", path.display());
+        let (snd, ret) = oneshot::channel();
+        self.sender
+            .send(Box::new(move |conn| {
+                let ret = perform_dump(conn, path);
+                let _ = snd.send(ret);
+            }))
+            .await
+            .map_err(|_| anyhow!("dump loader channel closed"))?;
```

Comment:
If the dump loading fails, will this message be all we see?
If that is the case, can we expose the underlying error with a message more meaningful to users?
Something like `loading DB from dump failed: <underlying error here>`.

Reply:
this error is unreachable in reality, the only exit condition to the loop is that we dropped the last sender, but we're currently holding one. I can't unwrap because the closure does not implement debug, so I just mock returning an error